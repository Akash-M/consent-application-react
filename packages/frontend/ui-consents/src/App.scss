@import "lib-components/src/scss/main";

main {
  align-items: flex-start;
  display: flex;
  flex-flow: column nowrap;
  height: 100vh;
  width: 100%;
}
