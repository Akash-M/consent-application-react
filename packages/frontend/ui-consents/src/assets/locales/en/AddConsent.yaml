header: Give consent
username: Name
email: Email Address
disclaimer: 'I agree to:'
addButton: Give consent
success: Consent added successfully
errors:
  username: Please enter a valid name
  email: Please enter a valid email id
  consents: Please ensure that you agree to at least one consent
  api: Error fetching consent list. Please try again in some time.
