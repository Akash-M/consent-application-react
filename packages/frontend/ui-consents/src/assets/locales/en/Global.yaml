headers:
  addConsent: Give consent
  listConsent: Collected consents
consents:
  newsletter: Receive newsletter
  ads: Be shown targeted ads
  statistics: Contribute to anonymous visit statistics
loader: Loading...
