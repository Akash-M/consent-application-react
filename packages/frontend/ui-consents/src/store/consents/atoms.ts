import { atom } from 'recoil';

// TODO: move this to `lib-api` and load from there.
const defaultConsentList: Consent.Detail[] = [
  {
    consentUuid: '2bd8e0da-c846-4d14-94b3-497638492175',
    username: 'asd',
    email: 'asd@gmail.com',
    consent: {
      newsletter: true,
      ads: false,
      statistics: true,
    },
  },
  {
    consentUuid: '894753bd-28bc-4b8e-884c-fcbae468e636',
    username: 'qwe',
    email: 'qwe@gmail.com',
    consent: {
      newsletter: true,
      ads: true,
      statistics: true,
    },
  },
  {
    consentUuid: 'bddd88d9-ca60-4bec-b1b6-55d0583aa619',
    username: 'zxc',
    email: 'zxc@gmail.com',
    consent: {
      newsletter: false,
      ads: false,
      statistics: false,
    },
  },
];

export const ConsentListState = atom<Consent.Detail[]>({
  key: 'ConsentList',
  default: defaultConsentList,
});

export interface ConsentListPaginator {
  perPage: number;
  currentPage: number;
}
export const ConsentListPaginatorState = atom<ConsentListPaginator>({
  key: 'ConsentListPaginator',
  default: {
    currentPage: 1,
    perPage: 2,
  },
});
