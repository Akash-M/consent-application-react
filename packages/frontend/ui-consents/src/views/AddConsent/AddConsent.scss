.add-consent {
  display: flex;
  flex-direction: column;
  margin: 1rem;

  @include tablet {
    width: auto;
  }

  @include desktop {
    margin: 0;
    max-width: 1024px;
    width: auto;
  }

  &__user-info {
    display: flex;
    flex-direction: column;
    margin-bottom: 2rem;

    .text-field {
      height: 40px;
    }

    .text-field:first-child {
      margin-bottom: 1rem;
    }

    @include tablet {
      flex-direction: row;

      .text-field {
        min-width: 250px;
      }

      .text-field:first-child {
        margin-bottom: 0;
      }

      .text-field:last-child {
        margin-left: 1rem;
      }
    }
  }

  &__consents {
    display: flex;
    flex-direction: column;
    margin-bottom: 2rem;
  }

  > button {
    align-self: flex-start;
    width: fit-content;
  }
}
