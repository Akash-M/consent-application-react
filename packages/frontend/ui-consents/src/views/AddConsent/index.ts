export { AddConsent as default } from '$/views/AddConsent/AddConsent';
