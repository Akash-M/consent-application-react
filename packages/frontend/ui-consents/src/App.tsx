import React from 'react';
import { ToastContainer } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import '$/App.scss';
import SiteHeader from '$/components/SiteHeader';
import Layout from '$/containers/Layout';

const App = (): JSX.Element => {
  return (
    <main>
      <SiteHeader />

      <Layout />

      <ToastContainer />
    </main>
  );
};

export default App;
