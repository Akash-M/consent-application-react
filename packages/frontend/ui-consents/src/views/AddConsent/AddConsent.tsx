import Button from '@mui/material/Button';
import Checkbox from '@mui/material/Checkbox';
import FormControl from '@mui/material/FormControl';
import FormControlLabel from '@mui/material/FormControlLabel';
import FormGroup from '@mui/material/FormGroup';
import FormHelperText from '@mui/material/FormHelperText';
import FormLabel from '@mui/material/FormLabel';
import TextField from '@mui/material/TextField';
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { toast } from 'react-toastify';

import { postConsent } from 'lib-api/src/consent';
import { EMAIL_REGEX } from 'lib-utils/src/email-regex';

import './AddConsent.scss';

export function AddConsent(): JSX.Element {
  const { t } = useTranslation(['AddConsent']);

  const [formData, setFormData] = useState<Consent.NewEntry>({
    username: '',
    email: '',
    consent: {
      newsletter: false,
      ads: false,
      statistics: false,
    },
  });
  const [isFormValid, setIsFormValid] = useState(true);

  const areConsentsValid = () => {
    return (
      formData.consent.newsletter ||
      formData.consent.ads ||
      formData.consent.statistics
    );
  };

  const updateConsent = (event: React.BaseSyntheticEvent) => {
    setFormData({
      username: formData.username,
      email: formData.email,
      consent: {
        ...formData.consent,
        [event.target.name]: event.target.checked,
      },
    });
  };

  const validateForm = () => {
    if (
      formData.username.length > 0 &&
      EMAIL_REGEX.test(formData.email) &&
      areConsentsValid()
    ) {
      setIsFormValid(true);
    } else {
      setIsFormValid(false);
    }
  };

  const handleSubmit = async (e: React.BaseSyntheticEvent) => {
    e.preventDefault();
    /* istanbul ignore else */
    if (isFormValid) {
      try {
        await postConsent(formData);
        toast.success(t('AddConsent.success'), {
          position: toast.POSITION.BOTTOM_RIGHT,
        });
      } catch {
        toast.error(t('AddConsent.errors.api'), {
          position: toast.POSITION.BOTTOM_RIGHT,
        });
      }
    }
  };

  return (
    <article className="add-consent">
      <h2>{t('AddConsent.header')}</h2>

      <form noValidate onSubmit={handleSubmit}>
        <section className="add-consent__user-info">
          <TextField
            required
            className="text-field"
            defaultValue={formData.username}
            error={!isFormValid && formData.username.length === 0}
            label={t('AddConsent.username')}
            name="username"
            variant="standard"
            onChange={(e) =>
              setFormData({ ...formData, username: e.target.value })
            }
          />

          <TextField
            required
            className="text-field"
            defaultValue={formData.email}
            error={!isFormValid && !EMAIL_REGEX.test(formData.email)}
            label={t('AddConsent.email')}
            name="email"
            variant="standard"
            onChange={(e) =>
              setFormData({ ...formData, email: e.target.value })
            }
          />
        </section>

        <section className="add-consent__consents">
          <FormControl
            required
            component="fieldset"
            error={!isFormValid && !areConsentsValid()}
            variant="standard"
          >
            <FormLabel component="legend">
              {t('AddConsent.disclaimer')}
            </FormLabel>

            <FormGroup>
              <FormControlLabel
                control={
                  <Checkbox
                    checked={formData.consent.newsletter}
                    name="newsletter"
                    onChange={updateConsent}
                  />
                }
                label={t<string>('Global.consents.newsletter')}
              />

              <FormControlLabel
                control={
                  <Checkbox
                    checked={formData.consent.ads}
                    name="ads"
                    onChange={updateConsent}
                  />
                }
                label={t<string>('Global.consents.ads')}
              />

              <FormControlLabel
                control={
                  <Checkbox
                    checked={formData.consent.statistics}
                    name="statistics"
                    onChange={updateConsent}
                  />
                }
                label={t<string>('Global.consents.statistics')}
              />
            </FormGroup>

            {!isFormValid && (
              <FormHelperText>{t('AddConsent.errors.consents')}</FormHelperText>
            )}
          </FormControl>
        </section>

        <Button type="submit" variant="contained" onClick={validateForm}>
          {t('AddConsent.addButton')}
        </Button>
      </form>
    </article>
  );
}
