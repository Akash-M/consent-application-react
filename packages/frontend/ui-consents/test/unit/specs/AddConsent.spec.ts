import { fireEvent, screen, waitFor } from '@testing-library/react';
import { setI18n } from 'react-i18next';

import { customRenderer } from 'lib-utils/src/testing/factory';
import { I18N_MISSING_KEY, loadI18n } from 'lib-utils/src/testing/i18n';

import App from '$/App';
import AddConsent from '$/assets/locales/en/AddConsent.yaml';
import Global from '$/assets/locales/en/Global.yaml';
import { ConsentListState } from '$/store/consents/atoms';
import { mockConsentList } from '#/fixtures/consent-list';

const initializeState = ({ set }: any) => {
  set(ConsentListState, mockConsentList);
};

describe('<App />', () => {
  beforeAll(() => {
    setI18n(loadI18n('Global', { Global, AddConsent }));
  });

  beforeEach(jest.clearAllMocks);

  test('should render screen to add new consent entry by default with button disabled', () => {
    const { container } = customRenderer(App, initializeState);
    expect(container.firstChild!.textContent).not.toContain(I18N_MISSING_KEY);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should append new entry to existing list and display toast message on success', async () => {
    customRenderer(App, initializeState);
    const inputs = screen.getAllByRole('textbox');

    fireEvent.change(inputs[0], { target: { value: 'user4' } });
    fireEvent.change(inputs[1], { target: { value: 'user4@email.com' } });
    fireEvent.click(screen.getByText('Global.consents.newsletter'));
    fireEvent.click(screen.getByText('Global.consents.ads'));
    fireEvent.click(screen.getByText('Global.consents.statistics'));
    fireEvent.click(screen.getByRole('button'));
    await waitFor(() => {
      expect(screen.getByText('AddConsent.success')).toBeTruthy();
    });
  });
});
