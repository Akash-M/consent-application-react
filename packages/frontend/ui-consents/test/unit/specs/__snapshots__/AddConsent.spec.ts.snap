// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render screen to add new consent entry by default with button disabled 1`] = `
<main>
  <nav
    class="site-header"
  >
    <h2>
      Consents App
    </h2>
  </nav>
  <article
    class="layout"
  >
    <aside
      class="layout__links"
    >
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        Global.headers.addConsent
      </a>
      <a
        class=""
        href="/list-consent"
      >
        Global.headers.listConsent
      </a>
    </aside>
    <section
      class="layout__screens"
    >
      <article
        class="add-consent"
      >
        <h2>
          AddConsent.header
        </h2>
        <section
          class="add-consent__user-info"
        >
          <div
            class="MuiFormControl-root MuiTextField-root text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              AddConsent.username
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="true"
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                name="username"
                required=""
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
            >
              AddConsent.email
              <span
                aria-hidden="true"
                class="MuiInputLabel-asterisk MuiFormLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="true"
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                name="email"
                required=""
                type="text"
                value=""
              />
            </div>
          </div>
        </section>
        <section
          class="add-consent__consents"
        >
          <fieldset
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <legend
              class="MuiFormLabel-root MuiFormLabel-colorPrimary Mui-error Mui-required css-u4tvz2-MuiFormLabel-root"
            >
              AddConsent.disclaimer
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
              >
                 
                *
              </span>
            </legend>
            <div
              class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
            >
              <label
                class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
              >
                <span
                  class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    name="newsletter"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                >
                  Global.consents.newsletter
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
              >
                <span
                  class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    name="ads"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                >
                  Global.consents.ads
                </span>
              </label>
              <label
                class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
              >
                <span
                  class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    name="statistics"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <span
                  class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
                >
                  Global.consents.statistics
                </span>
              </label>
            </div>
            <p
              class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium Mui-required css-1d1r5q-MuiFormHelperText-root"
            >
              AddConsent.errors.consents
            </p>
          </fieldset>
        </section>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          AddConsent.addButton
        </button>
      </article>
    </section>
  </article>
  <div
    class="Toastify"
  />
</main>
`;
